Add devicetree-based MMU region configuration

kedareswararao opened the pull request:
This pull request does the below
1) Enable automatic MMU configuration for ARM64 peripherals using devicetree
memory attributes. This eliminates the need to manually maintain region
lists in code when adding new peripherals.
2) Add test case for testing device-tree based memory attributes for ARM64 platform 

Tested changes on qemu(A53 and Versal Net) and Versalnet APU hardware

JiafeiPan reviewed (dismissed):
I once submitted one PR to achieve similar function, according to review comments, we still depend on device driver to take care of memory mapping.
https://github.com/zephyrproject-rtos/zephyr/pull/49908

Add it seems this PR need to add separate dts nodes but not reuse device's dts nodes' "reg" property, it is not convenient.

mbolivar reviewed (commented):
No objections to the DT changes.

kedareswararao commented:
> No objections to the DT changes.

Thanks @mbolivar if you are okay with the changes could you please approve

kedareswararao commented:
@JiafeiPan @mbolivar @nashif : Please review when you have sometime

kedareswararao commented on tests/arch/arm64/arm64_mmu_mem_attr/README.rst:
@pdgendt : Thanks for the review. In the initial implementation, the memory attributes were defined directly within the peripheral nodes. Based on @JiafeiPan's feedback, the attribute handling has been separated into a dedicated node.

One alternative approach is to introduce a unique mmu-regions node that references peripherals via phandles, for example:

/ {
    mmu-regions {
        compatible = "zephyr,mmu-regions";
        #address-cells = <2>;
        #size-cells = <2>;
        
        uart0-mmu {
            device = <&uart0>;  /* phandle to get address/size */
            zephyr,memory-attr = <DT_MEM_ARM64_MMU_DEVICE>;
        };
    };
};

if this approach is okay will update the code changes accordingly, please let me know your thoughts

carlocaione reviewed (changes requested):
We have `MMU_REGION_DT_COMPAT_FOREACH_FLAT_ENTRY` already that it does almost this but not entirely.

What about something like this instead?

```c
#define MMU_REGION_DT_COMPAT_FOREACH_FLAT_ENTRY_FROM_DT(compat) \
      DT_FOREACH_STATUS_OKAY_VARGS(compat, MMU_REGION_DT_FLAT_ENTRY_FROM_DT)

  #define MMU_REGION_DT_FLAT_ENTRY_FROM_DT(node_id) \
      MMU_REGION_FLAT_ENTRY(DT_NODE_FULL_NAME(node_id), \
                            DT_REG_ADDR(node_id), \
                            DT_REG_SIZE(node_id), \
                            DT_MEM_ATTR_TO_MT(DT_PROP(node_id, zephyr_memory_attr))),

  static const struct arm_mmu_region mmu_regions[] = {
      MMU_REGION_FLAT_ENTRY("GIC", ...),
      MMU_REGION_FLAT_ENTRY("GIC", ...),
      /* Memory regions, auto from DT with per-node attributes */
      MMU_REGION_DT_COMPAT_FOREACH_FLAT_ENTRY_FROM_DT(zephyr_memory_region)
  };
```
So we would:
- have the generation at compile time instead of run-time
- reuse what is already there
- one single line in each `mmu_regions.c`

This is all untested so I'm possibly making this easier than it actually is.

JiafeiPan reviewed (commented):
Would you please update the document for how to use this feature?

mathieuchopstm reviewed (dismissed):
-1 for the enum style attributes - ARM32 is already enough of a mess due to that

mathieuchopstm reviewed (commented):
much better!

carlocaione reviewed (changes requested):
A couple of more comments but this is looking almost fine already.

Commit:
arch: arm64: mmu: add devicetree-based MMU region configuration

Add compile-time macros to auto-generate ARM64 MMU region entries
from devicetree nodes with zephyr,memory-attr property:

- DT_MEM_ATTR_TO_MT(): converts DT memory attributes to ARM64 MT_*
  flags at compile time using the composable-bitmask design — the
  generic DT_MEM_CACHEABLE bit selects cacheable vs non-cacheable,
  and the arch-specific ATTR_ARM64_CACHE_WB sub-bit selects
  write-back vs write-through when cacheable
- MMU_REGION_DT_FLAT_ENTRY_FROM_DT(): generates a single MMU region
  entry from a DT node, skipping nodes without zephyr,memory-attr
- MMU_REGION_DT_COMPAT_FOREACH_FLAT_ENTRY_FROM_DT(): iterates over
  all status="okay" nodes of a given compatible

The convenience macros in memory-attr-arm64.h are defined as
composable combinations of generic and arch-specific attributes
rather than standalone one-hot enumerations:

  DT_MEM_ARM64_MMU_NORMAL_NC = (0)
  DT_MEM_ARM64_MMU_NORMAL_WT = (DT_MEM_CACHEABLE)
  DT_MEM_ARM64_MMU_NORMAL    = (DT_MEM_CACHEABLE | ATTR_ARM64_CACHE_WB)

Only Normal memory types are supported by the DT auto-mapping
mechanism. Device memory must be mapped through the MMIO device
API instead. Unsupported zephyr,memory-attr values are rejected
at compile time via BUILD_ASSERT.

The arch-level mmu.c automatically picks up all zephyr,memory-region
nodes with zephyr,memory-attr at boot, so individual SoC
mmu_regions.c files do not need any changes.

The DT region iteration is factored into max_region_bounds() and
map_mmu_regions() helpers to avoid code duplication in
setup_page_tables().

Introduces include/zephyr/dt-bindings/memory-attr/memory-attr-arm64.h
with ARM64-specific DT memory attribute sub-bits.

Add an ARM64 Developer Guide (doc/hardware/arch/arm64.rst)
documenting the DT-based MMU region mapping feature.

Signed-off-by: Appana Durga Kedareswara rao <appana.durga.kedareswara.rao@amd.com>

Commit:
tests: lib: shared_multi_heap: fix ARM64 MMU attributes

The qemu_cortex_a53 overlay was using ARM MPU memory attributes
(DT_MEM_ARM(ATTR_MPU_RAM)) even though Cortex-A53 is an ARM64
platform with MMU, not MPU.

Fix by switching the overlay to DT_MEM_ARM64_MMU_NORMAL /
DT_MEM_ARM64_MMU_NORMAL_NC and updating the test source to
handle both ARM MPU and ARM64 MMU attributes via
#if defined(CONFIG_ARM64).

The ARM64 path now checks the generic DT_MEM_CACHEABLE bit
instead of switching on architecture-specific enum values,
consistent with the composable-bitmask design.

Use the generic DT_MEM_ARCH_ATTR_UNKNOWN for the "unknown
attribute" default instead of architecture-specific aliases.

Remove the runtime smh_reg_map() function and its call site
since memory regions are now statically mapped at compile time
via MMU_REGION_DT_COMPAT_FOREACH_FLAT_ENTRY_FROM_DT, making
the dynamic k_mem_map_phys_bare() mapping redundant.

Increase CONFIG_MAX_XLAT_TABLES to 16 for qemu_cortex_a53.

Signed-off-by: Appana Durga Kedareswara rao <appana.durga.kedareswara.rao@amd.com>

kedareswararao commented on include/zephyr/arch/arm64/arm_mmu.h:
```diff
@@ -191,6 +193,67 @@ struct arm_mmu_ptables {
+		((_DT_MEM_ARM64_ARCH_BITS(dt_attr) & ATTR_ARM64_CACHE_WB) ?	\
+			(MT_NORMAL | MT_P_RW_U_NA | MT_DEFAULT_SECURE_STATE) :	\
+			(MT_NORMAL_WT | MT_P_RW_U_NA | MT_DEFAULT_SECURE_STATE)) : \
+		(MT_NORMAL_NC | MT_P_RW_U_NA | MT_DEFAULT_SECURE_STATE))
```
> The only problem with this is that now the unsupported `zephyr,memory-attr` values are silently mapped as normal memory instead of being rejected.
> 
> Can we have something like:
> 
> ```c
> #define ARM64_MMU_VALIDATE_DT_REGION(node_id)                                  \
>        IF_ENABLED(DT_NODE_HAS_PROP(node_id, zephyr_memory_attr),               \
>                (BUILD_ASSERT(DT_MEM_ARM64_MMU_IS_VALID(                        \
>                        DT_PROP(node_id, zephyr_memory_attr)),                  \
>                        "Unsupported zephyr,memory-attr for ARM64 MMU region");))
> 
> #define DT_MEM_ARM64_MMU_IS_VALID(dt_attr) ...
> 
> DT_FOREACH_STATUS_OKAY(zephyr_memory_region, ARM64_MMU_VALIDATE_DT_REGION)
> ```

Thanks for the review updated now

carlocaione reviewed (approved):
This is nice, thanks for the effort.

mbolivar reviewed (changes requested):
Putting a -1 since someone was asking for a merge on this. I want to review when the merge window opens again. Thanks.

mbolivar reviewed (approved):
Please don't define new `DT_` macros like this outside of `include/zephyr/devicetree*`. I see the namespace has gotten pretty messy and I'm working on cleaning it up, but in the meantime please don't make it worse.

I'm writing up some docs on how I usually review PRs like this, but the TL;DR for now is:

- the `DT_` namespace belongs to `<zephyr/devicetree.h>`
